data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,562 @@
+
+Execution evaluates each expression from top to bottom.
+
+At any time during evaluation, the _environment_ is the ordered sequence of available definitions.
+
+Here, the name `tau` is added into the toplevel environment.
+```ocaml
+# let tau = 6.28318;;
+val tau : float = 6.28318
+```
+
+The scope of `tau` is global. This name is available anywhere after the definition.
+
+Here, the global environment is unchanged:
+```ocaml
+# let pi = 3.14159 in 2. *. pi;;
+- : float = 6.28318
+```
+
+However, with respect to the expression `2. *. pi`, the environment is different, as it contains the definition of `pi`. Local definition create local environments.
+
+The scope of a definition is the set of environments where it is reachable.
+
+Although OCaml is an expression-oriented language, it is not entirely free of statements. The `let` construct is a statement that may change the state of the environment by adding a name-value binding. Note these expressions at the toplevel also fall into that category because they are equivalent to `let _ =` definitions.
+```ocaml
+# (1.0 +. sqrt 5.0) /. 2.0;;
+- : float = 1.6180339887498949
+
+# let _ = (1.0 +. sqrt 5.0) /. 2.0;;
+- : float = 1.6180339887498949
+```
+
+With respect to the environment, there are no means to:
+- List its contents
+- Remove a definition
+- Reset it to an earlier state
+
+#### Same-Level Shadowing
+
+You can write two definitions using the same identifier in sequence:
+```ocaml
+# let c = 2 * 3;;
+val c : int = 6
+
+# let d = c * 7;;
+val d : int = 42
+
+# let c = 7 * 7;;
+- : int = 49
+
+# d;;
+- : int = 42
+```
+
+In this example, `c` is defined twice. The key thing to understand is that `c` is not updated. It looks as if the value of `c` has changed, but it hasn't. When the second definition takes place, the first one becomes unreachable, but it still remains in the environment.
+
+Here is how this can be useful:
+```ocaml
+# let rec length len = function
+  | [] -> len
+  | _ :: u -> length (len + 1) u;;
+val length : int -> 'a list -> int = <fun>
+
+# let length u = length 0 u;;
+val length : 'a list -> int = <fun>
+```
+
+The first definition of `length` has two parameters: the list to process and the accumulated length that has already been computed. Writing it this way makes it tail-recursive.
+
+In practice, computing a list's length always starts with the accumulator set to nil. Therefore, the first definition is shadowed by the second where the accumulator parameter is set to 0.
+
+#### Inner Shadowing
+
+A local definition may shadow any other definition, just like a global definition does. The only difference is that shadowing is limited to the local definition's scope.
+```ocaml
+# let d = 21;;
+val d : int = 21
+
+# let d = 7 in d * 2;;
+- : int = 14
+
+# d;;
+- : int = 21
+```
+
+## Function Values
+
+### Restrictions
+
+Functions are supposed to be exactly as other values; however, there are three restrictions.
```

Comment:
If I'm reading this correctly, the three restrictions are:
1. Function values cannot be displayed in interactive sessions
2. Equality between functions can't be tested
3. Pattern matching does not allow users to inspect a function

Do you think these should be numbered or bolded to stand out and make clear these are the three? Especially since there is another list in between 2 and 3.